File: coderbyte/combination.py
def combinations(arr):
    if len(arr) == 0: return [[]]
    first_element = arr[0]
    rest = arr[1:]

    combs_without_first = combinations(rest)
    combs_with_first = []

    for x in combs_without_first:
        comb_with_first = [*x, first_element]
        combs_with_first.append(comb_with_first)
        print(x, comb_with_first, combs_with_first, sep=' # ')

    return [*combs_with_first, *combs_without_first]


print(combinations(['a', 'b', 'c']))

File: coderbyte/combination.js
const combinations = (elements) => {
    if (elements.length === 0) return [ []];
    const firstEl = elements[0];
    const rest = elements.slice(1);

    const combsWithoutFirst = combinations(rest);
    const combsWithFirst = [];

    combsWithoutFirst.forEach( comb => {
        const combWithFirst = [...comb, firstEl];
        combsWithFirst.push(combWithFirst)
        console.log(comb, combWithFirst, combsWithFirst)
    });

    return [...combsWithoutFirst, ...combsWithFirst]
}

console.log(combinations(['a', 'b', 'c']))
